modules/tracing/mule-tracer-exporter-impl/src/main/java/org/mule/runtime/tracer/impl/exporter/OpenTelemetrySpanExporter.java

```diff
@@ -7,94 +7,222 @@
+
+    // Generates the span id so that the opentelemetry spans can be lazily initialised.
+    this.spanContext = SpanContext.create(generateTraceId(getAsInternalSpan(internalSpan.getParent())), generateSpanId(),
+                                          TraceFlags.getSampled(), TraceState.getDefault());
+  }
+
+  @Nullable
+  @Override
+  public <T> T get(@Nullable AttributeKey<T> attributeKey) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void forEach(BiConsumer<? super AttributeKey<?>, ? super Object> biConsumer) {
+    biConsumer.accept(ARTIFACT_ID, artifactId);
+    biConsumer.accept(ARTIFACT_TYPE, artifactType);
+    biConsumer.accept(THREAD_END_NAME_KEY, endThreadNameValue);
+    internalSpan.forEachAttribute((key, value) -> biConsumer.accept(stringKey(key), value));
+  }
+
+  @Override
+  public int size() {
+    return 3 + internalSpan.getAttributesCount();
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return false;
+  }
+
+  @Override
+  public Map<AttributeKey<?>, Object> asMap() {
+    Map<AttributeKey<?>, Object> attributes = new HashMap<>();
+    forEach(attributes::put);
+    return attributes;
+  }
+
+  public static OpenTelemetrySpanExportBuilder builder() {
+    return new OpenTelemetrySpanExportBuilder();
+  }
+
+  @Override
+  public AttributesBuilder toBuilder() {
+    throw new UnsupportedOperationException();
   }
 
   @Override
   public void export() {
-    openTelemetrySpan.end(internalSpan, initialSpanInfo, artifactId, artifactType);
+    if (exportable) {
+      endThreadNameValue = Thread.currentThread().getName();
+      spanProcessor.onEnd(this);
+    }
   }
 
   @Override
   public void updateNameForExport(String newName) {
     if (rootSpan != this) {
       rootSpan.updateNameForExport(newName);
     } else {
-      openTelemetrySpan.updateName(newName);
+      overriddenSpanName = newName;
     }
   }
 
   @Override
   public Map<String, String> exportedSpanAsMap() {
-    return openTelemetrySpan.getDistributedTraceContextMap();
-  }
-
-  public MuleOpenTelemetrySpan getOpenTelemetrySpan() {
-    return openTelemetrySpan;
+    return OpenTelemetryTraceIdUtils.getContext(this);
   }
 
   @Override
   public void setRootName(String rootName) {
-    if (openTelemetrySpan.isRoot()) {
-      openTelemetrySpan.updateName(rootName);
+    if (isRootSpan) {
+      overriddenSpanName = rootName;
     } else {
       this.rootName = rootName;
     }
   }
 
   @Override
   public void setRootAttribute(String rootAttributeKey, String rootAttributeValue) {
-    if (openTelemetrySpan.isRoot()) {
-      openTelemetrySpan.setAttribute(rootAttributeKey, rootAttributeValue);
+    if (isRootSpan) {
+      internalSpan.addAttribute(rootAttributeKey, rootAttributeValue);
     } else {
       this.rootAttributes.put(rootAttributeKey, rootAttributeValue);
     }
   }
 
+  @Override
+  public void updateParentSpanFrom(Map<String, String> serializeAsMap) {
```

Comment:
What does serializeAsMap represent? Is it a span serialized as a map? Can we change the name to "spanSerializedAsMap" or "mapSerializedSpan"?

Reply:
it is a map represented via W3C format. I have added it to the javadoc.